File: articles/perma-model-of-well-being/c/perma_indices.c
#include <stdio.h>

typedef struct {
    double positive_emotion;
    double engagement;
    double relationships;
    double meaning;
    double accomplishment;
    double flourishing_score;
    double life_satisfaction;
    double institutional_support;
    double institutional_barriers;
    double autonomy_support;
    double fairness_score;
    double psychological_safety;
    double access_score;
    double workload_strain;
} PermaIndicators;

double perma_index(PermaIndicators x) {
    return (x.positive_emotion + x.engagement + x.relationships + x.meaning + x.accomplishment) / 5.0;
}

double institutional_quality(PermaIndicators x) {
    return x.institutional_support + x.autonomy_support + x.fairness_score +
           x.psychological_safety + x.access_score -
           x.institutional_barriers - x.workload_strain;
}

double context_adjusted_flourishing(PermaIndicators x) {
    return x.flourishing_score + x.life_satisfaction + perma_index(x) + institutional_quality(x);
}

int main(void) {
    PermaIndicators example = {6.9, 7.1, 7.2, 7.0, 7.1, 7.1, 7.0, 7.2, 3.1, 7.1, 7.0, 7.2, 7.1, 3.4};

    printf("PERMA index: %.3f\n", perma_index(example));
    printf("Institutional quality: %.3f\n", institutional_quality(example));
    printf("Context-adjusted flourishing: %.3f\n", context_adjusted_flourishing(example));

    return 0;
}

File: articles/perma-model-of-well-being/rust/perma_indices.rs
struct PermaIndicators {
    positive_emotion: f64,
    engagement: f64,
    relationships: f64,
    meaning: f64,
    accomplishment: f64,
    flourishing_score: f64,
    life_satisfaction: f64,
    institutional_support: f64,
    institutional_barriers: f64,
    autonomy_support: f64,
    fairness_score: f64,
    psychological_safety: f64,
    access_score: f64,
    workload_strain: f64,
}

fn perma_index(x: &PermaIndicators) -> f64 {
    (x.positive_emotion + x.engagement + x.relationships + x.meaning + x.accomplishment) / 5.0
}

fn institutional_quality(x: &PermaIndicators) -> f64 {
    x.institutional_support
        + x.autonomy_support
        + x.fairness_score
        + x.psychological_safety
        + x.access_score
        - x.institutional_barriers
        - x.workload_strain
}

fn context_adjusted_flourishing(x: &PermaIndicators) -> f64 {
    x.flourishing_score + x.life_satisfaction + perma_index(x) + institutional_quality(x)
}

fn main() {
    let example = PermaIndicators {
        positive_emotion: 6.9,
        engagement: 7.1,
        relationships: 7.2,
        meaning: 7.0,
        accomplishment: 7.1,
        flourishing_score: 7.1,
        life_satisfaction: 7.0,
        institutional_support: 7.2,
        institutional_barriers: 3.1,
        autonomy_support: 7.1,
        fairness_score: 7.0,
        psychological_safety: 7.2,
        access_score: 7.1,
        workload_strain: 3.4,
    };

    println!("PERMA index: {:.3}", perma_index(&example));
    println!("Institutional quality: {:.3}", institutional_quality(&example));
    println!("Context-adjusted flourishing: {:.3}", context_adjusted_flourishing(&example));
}
